### PR TITLE
Add support for objects and classes

### PR DESCRIPTION
**Pull request by hhugo**
This is an early preview. Feedback and contributions are welcome

### CODE REVIEW

**Comment by jberdine** [comment]
Excellent, looking great! I couldn't check in great detail, but reading through the implementation I did not see anything concerning or notice any lurking gotchas.

**Comment by hhugo** [comment]
Certainly not perfect, not bug free, but I think I'm done the initial support.

File: src/Fmt_ast.ml [hunk 1]
```diff
@@ -1730,8 +1781,400 @@ and fmt_expression c ?(box= true) ?epi ?eol ?parens ?ext
...
+        hvbox_if box 2
+          (fmt_label lbl ":@," $ fmt_expression c ~box ?epi ?parens xarg)
+  in
+  let ctx = Cl exp in
+  let width xe =
+    String.length
+      (Format.asprintf "%t" (Cmts.preserve (fun () -> fmt_expression c xe)))
+  in
+  let fmt_args_grouped e0 a1N =
+    (* TODO: consider [e0] when groupping *)
```
Comment (jberdine) [inline src/Fmt_ast.ml]: nit: s/groupping/grouping .

This is nontrivial due to the different types of e0 and a1N, right? Any other reason?

Comment (hhugo) [reply src/Fmt_ast.ml]: no other reason.

**Comment by jberdine** [comment]
Thanks!

**Pull request merged**